src/agent/transform-headers-agent.ts

```diff
@@ -67,11 +67,10 @@ class TransformHeadersAgent extends WrappedAgent {
             }
         }
 
-        const transformedHeaders = sortHeaders ? generator.orderHeaders(headers) : headers;
+        const transformedHeaders: Record<string, string | number | string[]> = sortHeaders ? generator.orderHeaders(headers) : headers;
 
-        // eslint-disable-next-line no-restricted-syntax, guard-for-in
-        for (const key in transformedHeaders) {
-            request.setHeader(key, transformedHeaders[key]);
+        for (const [key, value] of Object.entries(transformedHeaders)) {
+            request.setHeader(key, value);
```

Comment:
@vladfrangu This is exactly slower and is not stricter (nor loose) because `transformedHeaders` has cloned entries.

Reply:
A for-in loop cannot be faster than a for-of, and it's far more risky due to prototype pollution, and it's stricter in the context of TS (and it picking just the object's own properties)

Reply:
80% slower

<details>

<summary>source</summary>

https://jsbench.me/

#### setup

```js
const o = {
	foo: 'bar',
	hello: 'world',
	connection: 'keep-alive',
	host: 'example.com',
	'content-length': '16',
	'accept-encoding': 'gzip',
	'user-agent': 'lorem impsum'
};
```

#### 1

```js
for (let key in o) {
	o[key];
}
```

#### 2

```js
for (const [key, value] of Object.entries(o)) {}
```

</details>

Reply:
Touche. For an ideal world where there's no prototype pollution, for-in might be faster, but not safer. Consider this:
![image](https://user-images.githubusercontent.com/17960496/129175157-202bf74f-db1e-47a8-9b22-84839ff4b436.png)

Reply:
> A for-in loop cannot be faster than a for-of

I'd love to see an example :) Haven't really met this case, but I don't negate it.

> it's far more risky due to prototype pollution

No prototype pollution here, the entries are cloned before.

> it's stricter in the context of TS (and it picking just the object's own properties)

You're right. I'll leave this up to you then :)

Reply:
> No prototype pollution here, the entries are cloned before.

*We* don't have prototype pollutions 😄 
Other dependencies, user code, etc might tho. Thats where the speed balance shifts 😅

Reply:
Prototype mutation on the native types should be removed from V8 IMO. It's strongly discouraged to do this. The entire app's behavior can change. Node.js circumvents this via primodals.

Reply:
> Prototype mutation on the native types should be removed from V8 IMO. It's strongly discouraged to do this. The entire app's behavior can change. Node.js circumvents this via primodals.

Oh I am aware and I strongly agree, but till that day comes, I think the best compromise we have right now for speed/safety is Object.keys. What do you think?

Reply:
Dunno. `Object.keys` can be overwritten as well (unlikely tho) 😅 I'm good with anything that works :D

Reply:
CC @mnmkng thoughts? ^

Reply:
I tried to find a nice article I read, but it said something like this: The lengthiest and most heated discussions in code reviews are spent on discussing theoretical problems and code preferences of individual developers 😄 

I personally don't use `for in`. Not because I think it's always bad, but because it makes people argue and it's just not worth arguing about. We have about 15 other ways of writing a loop which do not promote heated discussions ❤️

The performance here is irrelevant really. For each request that will take on average hundreds of milliseconds, we don't really care if it the client spends 1 microsecond or 2 microseconds in a loop.

So let's just keep whatever loop is in there now 🤝